File: app/src/main/java/com/jack/autostart/ui/MainActivity.java
package com.jack.autostart.ui;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.LayoutInflater;
import android.widget.Toast;

import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;

import com.jack.autostart.app.BaseApplication;
import com.jack.autostart.databinding.MainActivityBinding;
import com.jack.autostart.ui.list.AppInfoListItemAdapter;
import com.jack.autostart.ui.model.AppInfo;
import com.jack.autostart.ui.model.AppInfoListViewModel;
import com.jack.autostart.utils.AppLauncherUtils;
import com.jack.autostart.utils.handler.MyHandlerThread;

import java.util.Comparator;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private MainActivityBinding mBinding;
    private AppInfoListViewModel mViewModel;
    private AppInfoListItemAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d(TAG, "onCreate");
        mBinding = MainActivityBinding.inflate(LayoutInflater.from(MainActivity.this));
        setContentView(mBinding.getRoot());
        initViewModel();
        initView();
        checkDrawOverlaysPermission();
    }

    private void initViewModel() {
        mViewModel = new ViewModelProvider(this).get(AppInfoListViewModel.class);
        mViewModel.fetchAllInstalledApps();
    }

    private void initView() {
        mAdapter = new AppInfoListItemAdapter(mViewModel.getAppInfos(), mOnItemListener);
        mBinding.appList.setAdapter(mAdapter);
        mBinding.appList.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
    }

    private final AppInfoListItemAdapter.OnItemListener mOnItemListener = new AppInfoListItemAdapter.OnItemListener() {
        @Override
        public void onItemClick(AppInfo appInfo) {
            mViewModel.addSelectItem(appInfo);
            mAdapter.update();
        }

        @Override
        public void onItemLongClick(AppInfo appInfo) {
            mViewModel.removeSelectItem(appInfo);
            mAdapter.update();
        }

        @Override
        public void onMinusClick(AppInfo appInfo) {
            mViewModel.minus(appInfo);
            mAdapter.update();
        }

        @Override
        public void onPlusClick(AppInfo appInfo) {
            mViewModel.plus(appInfo);
            mAdapter.update();
        }
    };

    private void checkDrawOverlaysPermission() {
        if (!Settings.canDrawOverlays(BaseApplication.getContext())) {
            // Initialize ActivityResultLauncher for overlay permission
            Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION);
            intent.setData(Uri.parse("package:" + BaseApplication.getContext().getPackageName()));
            registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
                if (Settings.canDrawOverlays(BaseApplication.getContext())) {
                    startLaunchApp();
                } else {
                    Toast.makeText(BaseApplication.getContext(), "Overlay permission denied", Toast.LENGTH_LONG).show();
                }
            }).launch(intent);
        } else {
            startLaunchApp();
        }
    }

    private void startLaunchApp() {
        List<AppInfo> selectedAppsInfo = mViewModel.getSelectedAppsInfo();
        selectedAppsInfo.sort(Comparator.comparingInt(AppInfo::getOrder));
        MyHandlerThread handler = new MyHandlerThread();
        long totalDelay = 0;
        for (AppInfo appInfo : selectedAppsInfo) {
            long delayTime = appInfo.getDelaySec() * 1000L; // 將秒轉換為毫秒
            totalDelay += delayTime; // 累積延遲時間
            handler.postDelayed(() -> {
                AppLauncherUtils.launchAppWithPackageName(appInfo.getPackageName());
                Log.d(TAG, "Launching app: " + appInfo);
            }, totalDelay);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "onDestroy");
    }
}

File: app/src/main/java/com/jack/autostart/MainActivity.kt
package com.jack.autostart

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.Image
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Button
import androidx.compose.material3.Divider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment.Companion.CenterVertically
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.core.graphics.drawable.toBitmap
import androidx.core.graphics.drawable.toDrawable
import com.jack.autostart.AppListViewModel.AppInfo
import com.jack.autostart.app.BaseApplication
import com.jack.autostart.ui.theme.AutoStartTheme
import com.jack.autostart.utils.AppLauncherUtils
import com.jack.autostart.utils.OnActResultDelegateFragment
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

private val TAG = MainActivity::class.simpleName

class MainActivity : AppCompatActivity() {

    private val viewModel: AppListViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, "onCreate")
        setContent {
            AutoStartTheme {
                // A surface container using the 'background' color from the theme
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    viewModel.getAllInstalledApps()
                    AppsInfoScreen(
                        viewModel.appsInfo,
                        { viewModel.addSelectItem(it) },
                        { viewModel.removeSelectItem(it) })
                }
            }
        }
        checkDrawOverlaysPermission()
    }

    private fun checkDrawOverlaysPermission() {
        val context = BaseApplication.context
        if (!Settings.canDrawOverlays(context)) {
            val intent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION)
            intent.setData(Uri.parse("package:" + context.packageName))
            OnActResultDelegateFragment.request(supportFragmentManager)
                .startForResult(intent) {
                    if (Settings.canDrawOverlays(context)) {
                        startLaunchApp()
                    } else {
                        Toast.makeText(context, "Overlay permission denied", Toast.LENGTH_LONG).show()
                    }
                }
        } else {
            startLaunchApp()
        }
    }

    private fun startLaunchApp() {
        GlobalScope.launch {
            for (appInfo in viewModel.getSelectedAppsInfo()) {
                AppLauncherUtils.launchAppWithPackageName(appInfo.packageName)
                Log.d(TAG, "appInfo $appInfo")
                delay(5000)
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("MainActivity", "onDestroy")
    }
}

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun AppsInfoScreen(
    appsInfo: List<AppInfo>,
    onClick: (AppInfo) -> Unit,
    onLongClick: (AppInfo) -> Unit,
) {
    Column {
        if (appsInfo.isEmpty()) {
            Text(text = "Loading...")
        } else {
            LazyColumn {
                items(appsInfo) { appInfo ->
                    Column {
                        Row(Modifier
                            .fillMaxSize()
                            .combinedClickable(
                                onClick = {
                                    onClick(appInfo)
                                },
                                onLongClick = {
                                    onLongClick(appInfo)
                                }
                            )
                            .padding(10.dp)) {
                            val bitmap: ImageBitmap = try {
                                appInfo.icon!!.toBitmap().asImageBitmap()
                            } catch (_: Exception) {
                                ImageBitmap(100, 100)
                            }
                            Image(
                                bitmap = bitmap,
                                contentDescription = "",
                                modifier = Modifier
                                    .padding(end = 10.dp)
                                    .size(80.dp)
                            )
                            Column {
                                Text(text = appInfo.appName)
                                Text(text = appInfo.packageName)
                            }
                            Spacer(Modifier.weight(1f))
                            if (appInfo.order != 9999) {
                                Text(
                                    text = appInfo.order.toString(),
                                    modifier = Modifier.align(CenterVertically)
                                )
                            }
                        }
                        Divider()
                    }
                }
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
fun AppsInfoScreenPreview() {
    AutoStartTheme {
        AppsInfoScreen(
            listOf(
                AppInfo(
                    "Settings",
                    "com.setting",
                    icon = R.drawable.ic_launcher.toDrawable()
                ),
                AppInfo(
                    "Camera",
                    "com.google.camera",
                    icon = R.drawable.ic_launcher.toDrawable()
                )
            ),
            {},
            {})
    }
}